core/crdt/pncounter.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2022 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package crdt
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/fxamacker/cbor/v2"
+	dag "github.com/ipfs/boxo/ipld/merkledag"
+	ds "github.com/ipfs/go-datastore"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ugorji/go/codec"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/base"
+	"github.com/sourcenetwork/defradb/errors"
+)
+
+var (
+	// ensure types implements core interfaces
+	_ core.ReplicatedData = (*PNCounterRegister)(nil)
+	_ core.Delta          = (*PNCounterDelta)(nil)
+)
+
+// PNCounterDelta is a single delta operation for an PNCounterRegister
+type PNCounterDelta struct {
+	DocKey    []byte
+	FieldName string
+	Priority  uint64
+	// SchemaVersionID is the schema version datastore key at the time of commit.
+	//
+	// It can be used to identify the collection datastructure state at the time of commit.
+	SchemaVersionID string
+	Data            []byte
+	FieldValue      client.FieldValue `json:"-"` // should not be marshalled
```

Comment:
It's akin to the `SubDAGs` fo the composite CRDT. It holds a value that is useful for the merge and decode operations but that don't need to be marshalled. If we don't keep the field value in a format that is easily usable, then we need to marshal if when we create the Delta and unmarshal it when we need the value in the Merge function.

Reply:
The similarties for `SubDAGs` doesn't exactly hold. The `SubDAGs` links slice needs to be used in a couple of unique instances, because `CompositeCRDT` has unique semantics, compared to regular CRDTs.

In this instance `FieldValue` really doesn't need to exist, as you can just accept numerics as a parameter on the delta-mutator function, and the `Data` field on the `Delta` can be the same numerics type. If it used generics for example, it would roughly look like this:

```go
type PNCounterDelta[T Numeric] struct {
  Data T
  // ... omitting other fields
}

type PNCounter[T Numeric] struct {
  baseCRDT
}

func (pnc *PNCounter[T]) Increment(amount T) PNCounterDelta[T] {
  return PNCounterDelta{Data: amount}
}
```

Reply:
Whats the benefit for having the Counter data as bytes?

Reply:
This is my point, each CRDT is a beautiful and unique snowflake. Each should be looked at more or less in isolation, as they each have their own unique and complex semantics. Register deals with bytes, Counter deals with numerics.

Reply:
I was trying to keep it as similar to one another as possible to try to simplify the code. I think I tried to go a step too far. I'll change it to use a numeric type.

Reply:
✅